### PR TITLE
Index Writer Framework Refactor

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

New Version of Index Writer
Remove changes of Spark code.
Utilize the whole code logical in FileFormatWriter (previously known as WriterContainer)
This mechanism can prevent multiple tasks writing same file

## How was this patch tested?
unit test pass

### CODE REVIEW

**Comment by linhong-intel** [comment]
Disabled statistics manager since it seems unnecessary.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 1]
```diff
@@ -366,32 +367,29 @@ case class RefreshIndex(
...
-        keySchema,
-        i.name,
-        i.time,
-        indexType,
-        isAppend = true)
-
-      indexWriter.write(sparkSession = sparkSession,
-        queryExecution = queryExecution,
-        fileFormat = indexFileFormat,
+      val options = Map(
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: Suggest using CaseInsensitiveMap just like DataSource

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    keySchema: StructType) extends RecordWriter[Void, InternalRow] {
+
+  @transient private lazy val genericProjector = FromUnsafeProjection(keySchema)
+
+  private val multiHashMap = ArrayListMultimap.create[InternalRow, Int]()
+  private var recordCount: Int = 0
+
+  override def write(key: Void, value: InternalRow): Unit = {
+    val v = genericProjector(value).copy()
+    multiHashMap.put(v, recordCount)
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: It should be done later. This PR is only framework changes.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+  private def flushToFile(out: OutputStream): Unit = {
+
+    val statisticsManager = new StatisticsManager
+    statisticsManager.initialize(BitMapIndexType, keySchema, configuration)
+
+    // generate the bitset hashmap
+    val hashMap = new mutable.HashMap[InternalRow, BitSet]()
+
+    multiHashMap.asMap().asScala.foreach(kv => {
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: should be done later. only framework changes.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala (hunk 1 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: sure

**Comment by yueawang** [comment]
LGTM

**Comment by lee-lei** [comment]
It's a too big change. It seems to me that this change consists of three categories. The first is to change the common file format writer for Spark and apply it for OAP. The second is to refactor or rename the OAP specific code except index. The third is to just refactor the index writer framework code. It's better to split to the above three changes or more to easily read and track. Thanks.

**Comment by linhong-intel** [comment]
Thank @lee-lei.
I will upload small PRs in the future development. But I'd like to merge this one to avoid conflicts.

**Pull request merged**